### PR TITLE
fix: promql issues

### LINKED ISSUE #746: Some metrics have no metadata on /metadata API
### Which ZincObserve functionalities are the source of the bug?

metrics

### Is this a regression?

Yes

### Description

Some metrics have no metadata on /metadata API, we should add some default, because empty metadata will cause problem.

### Please provide a link to a minimal reproduction of the bug

_No response_

### Please provide the exception or error you saw

_No response_

### Please provide the version you discovered this bug in (check about page for version information)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by hengfeiyang**
- [x] metadata API response fix #746 
- [x] series API supports selector #748

### CODE REVIEW

File: src/handler/http/request/prom/mod.rs [hunk 1]
```diff
@@ -104,13 +102,24 @@ pub async fn remote_write(
         (status = 500, description="Failure", content_type = "application/json", body = HttpResponse),
     )
 )]
-#[route("/{org_id}/prometheus/api/v1/query", method = "GET", method = "POST")]
```
Comment (vvv) [inline src/handler/http/request/prom/mod.rs:107]: @hengfeiyang So `method = "GET", method = "POST"` attributes didn't work, did they?

File: src/service/metrics/prom.rs [hunk 2]
```diff
@@ -473,11 +474,12 @@ pub(crate) async fn get_metadata(
...
-                (
-                    schema.stream_name,
-                    get_metadata_object(&schema.schema).map_or_else(Vec::new, |obj| vec![obj]),
-                )
+            let metric_names = stream_schemas.into_iter().filter_map(|schema| {
+                if let Some(meta) = get_metadata_object(&schema.schema) {
+                    Some((schema.stream_name, vec![meta]))
+                } else {
+                    None
+                }
```
Comment (vvv) [inline src/service/metrics/prom.rs:482]:
```suggestion
                get_metadata_object(&schema.schema).map(|meta| (schema.stream_name, vec![meta]))
```

File: src/service/metrics/prom.rs [hunk 3]
```diff
@@ -534,9 +535,40 @@ pub(crate) async fn get_series(
...
+            }
+            match &mat.op {
+                MatchOp::Equal => {
+                    sql_where.push(format!("{} = '{}'", mat.name, mat.value));
+                }
+                MatchOp::NotEqual => {
+                    sql_where.push(format!("{} != '{}'", mat.name, mat.value));
+                }
+                MatchOp::Re(_re) => {
+                    sql_where.push(format!("re_match({}, '{}'", mat.name, mat.value));
```
Comment (vvv) [inline src/service/metrics/prom.rs:556]: Shouldn't there be a closing parenthesis?
```suggestion
                    sql_where.push(format!("re_match({}, '{}')", mat.name, mat.value));
```

File: src/service/metrics/prom.rs [hunk 4]
```diff
@@ -534,9 +535,40 @@ pub(crate) async fn get_series(
...
+                }
+                MatchOp::NotRe(_re) => {
+                    sql_where.push(format!("re_not_match({}, '{}'", mat.name, mat.value));
```
Comment (vvv) [inline src/service/metrics/prom.rs:559]: Also here?

File: src/service/promql/engine.rs [hunk 5]
```diff
@@ -250,15 +250,21 @@ impl QueryEngine {
             None => return Ok(vec![]),
         };
 
+        let end = self.start + (self.interval * self.time_window_idx); // 15s
```
Comment (vvv) [inline src/service/promql/engine.rs:253]: [nit, optional]
```suggestion
        let end = self.start + (self.interval * self.time_window_idx); // e.g., 15s
```

File: src/service/promql/engine.rs [hunk 6]
```diff
@@ -250,15 +250,21 @@ impl QueryEngine {
...
+        let start = end - self.lookback_delta; // 5m
```
Comment (vvv) [inline src/service/promql/engine.rs:254]: [nit, optional]
```suggestion
        let start = end - self.lookback_delta; // e.g., 5m
```

Rationale: it's not always `5m` and `15s`, AFAIU.

File: src/service/metrics/prom.rs (hunk 3 above)
Comment (vvv) [inline src/service/metrics/prom.rs:556]: [nit, optional] Consider pushing once, i.e.

```rust
            let s = match &mat.op {
                MatchOp::Equal => format!("{} = '{}'", mat.name, mat.value),
                MatchOp::NotEqual => format!("{} != '{}'", mat.name, mat.value),
                ...
            };
            sql_where.push(s);

```

**Review by vvv — APPROVED** [review]

**Review by vvv — APPROVED** [review]

File: src/handler/http/request/prom/mod.rs (hunk 1 above)
Comment (hengfeiyang) [reply src/handler/http/request/prom/mod.rs:107]: It works, but the params are different, for GET need fetch pramas from query, for POST need fetch params from From. So must split into two functions.

File: src/service/promql/engine.rs (hunk 6 above)
Comment (hengfeiyang) [reply src/service/promql/engine.rs:254]: just delete the comment

**Pull request merged**